Source/Tools/FEXLogServer/Main.cpp

```diff
@@ -0,0 +1,276 @@
+  const char *GetCharLevel(uint32_t Level) {
+    const char *CharLevel{nullptr};
+
+    switch (Level) {
+    case LogMan::NONE:
+      CharLevel = "NONE";
+      break;
+    case LogMan::ASSERT:
+      CharLevel = "ASSERT";
+      break;
+    case LogMan::ERROR:
+      CharLevel = "ERROR";
+      break;
+    case LogMan::DEBUG:
+      CharLevel = "DEBUG";
+      break;
+    case LogMan::INFO:
+      CharLevel = "Info";
+      break;
+    case LogMan::STDOUT:
+      CharLevel = "STDOUT";
+      break;
+    case LogMan::STDERR:
+      CharLevel = "STDERR";
+      break;
+    default:
+      CharLevel = "???";
+      break;
+    }
+    return CharLevel;
+  }
+}
+
+namespace CLI {
+  static void MsgHandler(int FD, uint64_t Timestamp, uint32_t PID, uint32_t TID, uint32_t Level, const char* Msg) {
+    auto CharLevel = Common::GetCharLevel(Level);
+    const auto Output = fmt::format("[{}][{}][{}.{}] {}\n", CharLevel, Timestamp, PID, TID, Msg);
+    write(STDERR_FILENO, Output.c_str(), Output.size());
+  }
+}
+
+namespace GUI {
+  static std::chrono::time_point<std::chrono::high_resolution_clock> GlobalTime{};
+
+  // Configuration
+  static size_t MaxLogLines = 1000;
+  static LogMan::DebugLevels LevelFilter = LogMan::DebugLevels::INFO;
+  bool AlsoCLI {true};
+
+  // PID information
+  struct PIDMapping {
+    std::unordered_set<uint32_t> FDs{};
+    std::unordered_set<uint32_t> TIDs{};
+  };
+  // PID -> FD/TID mapping
+  std::map<uint32_t, PIDMapping> PIDs{};
+  std::mutex PIDMappingMutex{};
+
+  // Log window
+  static int LogSelected{};
+
+  struct LogData {
+    uint64_t Timestamp;
+    uint32_t PID;
+    uint32_t TID;
+    uint32_t Level;
+    int FD;
+    std::string Log{};
+  };
+
+  std::mutex LogMutex{};
+  std::vector<LogData> LogLines{};
+
+  static void MsgHandler(int FD, uint64_t Timestamp, uint32_t PID, uint32_t TID, uint32_t Level, const char* Msg) {
+    if (Level > LevelFilter) {
+      // Skip this log
+      return;
+    }
+
+    {
+      std::unique_lock lk{PIDMappingMutex};
+      auto it = PIDs.find(PID);
+      if (it == PIDs.end()) {
+        it = PIDs.try_emplace(PID, PIDMapping {
+        }).first;
+      }
+
+      it->second.FDs.emplace(FD);
+      it->second.TIDs.emplace(TID);
+    }
+
+    if (AlsoCLI) {
+      CLI::MsgHandler(FD, Timestamp, PID, TID, Level, Msg);
+    }
+
+    std::unique_lock lk{LogMutex};
+    auto CharLevel = Common::GetCharLevel(Level);
+    LogLines.emplace_back(LogData {
+        Timestamp,
+        PID,
+        TID,
+        Level,
+        FD,
+        fmt::format("[{}][{}][{}.{}] {}\n", CharLevel, Timestamp, PID, TID, Msg),
+    });
+
+    if (LogLines.size() > MaxLogLines) {
+      size_t NumberToRemove = LogLines.size() - MaxLogLines;
+      LogLines.erase(LogLines.begin(), LogLines.begin()+NumberToRemove);
+    }
+    FEX::GUI::HadUpdate();
+  }
+
+  static void FDClosedHandler(int FD) {
+    std::unique_lock lk{PIDMappingMutex};
+    for (auto it = PIDs.begin(); it != PIDs.end(); ) {
```

Comment:
This seems like it can be a `std::erase_if`

Reply:
Resolved.